packages/flutter/lib/src/material/material_state_mixin.dart

```diff
@@ -0,0 +1,130 @@
+// Copyright 2021 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'package:flutter/material.dart';
+
+import '../../foundation.dart';
+
+/// Provides a flexible callback and consistent API for [State] objects to track
+/// which [MaterialState] values are in effect for their portion of the user
+/// interface.
+///
+/// This mixin does nothing by merely applying it to a [State] class, but is
+/// helpful when writing `build` methods that include child [InkWell],
+/// [GestureDetector], [MouseRegion], or [Focus] widgets. Instead of manually
+/// creating handlers for each type of user interaction, such [State] classes can
+/// instead provide a `ValueChanged<bool>` function and allow [MaterialStateMixin]
+/// to manage the set of active [MaterialState]s, and the calling of `setState()`
+/// as necessary.
+///
+/// Use of this mixin changes the implementation below, in Example 1, into the
+/// ensuing, much shorter implementation in Example 2. The two code samples are
+/// functionally equivalent.
+///
+/// Example 1: Tracking MaterialState values without MaterialStateMixin.
+/// ```
+/// class MyWidgetState extends State<MyWidget> {
+///
+///   final Set<MaterialState> materialStates;
+///
+///   bool get _focused => materialStates.contains(MaterialState.focused);
+///
+///   void _handleFocusChanged(bool value) {
+///     if (value != _focused) {
+///       materialStates.add(MaterialState.focused);
+///     } else {
+///       materialStates.remove(MaterialState.focused);
+///     }
+///   }
+///
+///   @override
+///   Widget build(BuildContext context) {
+///     return InkWell(
+///       onFocusChange: () => _handleFocusChanged,
+///       child: Container(
+///         color: widget.color.resolve(materialStates),
+///         child: widget.child,
+///       ),
+///     );
+///   }
+/// }
+/// ```
+///
+/// Example 2: Tracking MaterialState values WITH MaterialStateMixin.
+/// ```
+/// class MyWidgetState extends State<MyWidget> with MaterialStateMixin<MyWidget> {
+///
+///   @override
+///   Widget build(BuildContext context) {
+///     return InkWell(
+///       onFocusChange: () => updateMaterialState(MaterialState.focused),
+///       child: Container(
+///         color: widget.color.resolve(materialStates),
+///         child: widget.child,
+///       ),
+///     );
+///   }
+/// }
+/// ```
+mixin MaterialStateMixin<T extends StatefulWidget> on State<T> {
+
+  /// Managed set of active [MaterialState] values.
+  Set<MaterialState> materialStates = <MaterialState>{};
+
+  /// Callback factory which accepts a [MaterialState] value and returns a
+  /// closure to mutate `self.materialStates` and call `setState()`.
+  ValueChanged<bool> updateMaterialState(MaterialState key, {ValueChanged<bool>? callback}) {
+    return (bool value) {
+      if (materialStates.contains(key) == value)
```

Comment:
Do we need to invoke the `callback` even if this is true?

Reply:
This is a good question. My intuition was/is that we do not, but I obviously could be wrong. If an item is currently _pressed_ or _hovered_ or what have you, re-entering that same state should both be uninteresting and, even worse, a bug.

If there are any known edge cases that imply the above is both intended and correct, then I think the answer to your question would change to a strong _Yes_.

Reply:
If re-entering that same state is a bug, then we should assert here if that happens. I am not sure if it is, though. A button could be in the pressed state already (because a finger is touching it) and now a second finger comes down sending it again into the pressed state it is already in. I don't think this is a bug.

Maybe we can just document on the method that the callback only executes if a change in MaterialState occurs?

Reply:
Maybe we should rename it from `callback` to `onChanged` to make it clear when this fires.